x-pack/legacy/plugins/index_management/static/ui/components/mappings_editor/components/document_fields/fields/fields_list_item.tsx

```diff
@@ -0,0 +1,123 @@
+ */
+import React, { useMemo } from 'react';
+import { EuiButton, EuiButtonEmpty } from '@elastic/eui';
+
+import { useState, useDispatch } from '../../../mappings_state';
+import { FieldsList } from './fields_list';
+import { CreateField } from './create_field';
+import { DeleteFieldProvider } from './delete_field_provider';
+import { NormalizedField } from '../../../types';
+import { MAX_DEPTH_DEFAULT_EDITOR } from '../../../constants';
+import { validateUniqueName } from '../../../lib';
+
+interface Props {
+  field: NormalizedField;
+  treeDepth?: number;
+}
+
+const inlineStyle = {
+  borderBottom: '1px solid #ddd',
+  display: 'flex',
+  flexDirection: 'column' as 'column',
+};
+
+export const FieldsListItem = ({ field, treeDepth = 0 }: Props) => {
+  const dispatch = useDispatch();
+  const {
+    documentFields: { status, fieldToAddFieldTo },
+    fields: { byId, rootLevelFields },
+  } = useState();
+  const getField = (propId: string) => byId[propId];
+  const { id, source, childFields, hasChildFields, canHaveChildFields } = field;
+  const isAddFieldBtnDisabled = field.nestedDepth === MAX_DEPTH_DEFAULT_EDITOR - 1;
+
+  const uniqueNameValidator = useMemo(() => {
+    return validateUniqueName({ rootLevelFields, byId }, undefined, id);
+  }, [byId, rootLevelFields]);
+
+  const addField = () => {
+    dispatch({
+      type: 'documentField.createField',
+      value: id,
+    });
+  };
+
+  const editField = () => {
+    dispatch({
+      type: 'documentField.editField',
+      value: id,
+    });
+  };
+
+  const renderCreateField = () => {
+    if (status !== 'creatingField') {
+      return null;
+    }
+
+    // Root level (0) has does not have the "fieldToAddFieldTo" set
+    if (fieldToAddFieldTo !== id) {
+      return null;
+    }
+
+    return (
+      <div style={{ paddingLeft: '20px' }}>
+        <CreateField uniqueNameValidator={uniqueNameValidator} />
+      </div>
+    );
+  };
+
+  const renderActionButtons = () => {
+    if (status !== 'idle') {
+      return null;
+    }
+
+    return (
+      <>
+        <EuiButton onClick={editField}>Edit</EuiButton>
+        {canHaveChildFields && (
+          <>
+            <EuiButton onClick={addField} disabled={isAddFieldBtnDisabled}>
+              Add field
+            </EuiButton>
+          </>
+        )}
+        <DeleteFieldProvider>
+          {deleteField => <EuiButton onClick={() => deleteField(field)}>Remove</EuiButton>}
+        </DeleteFieldProvider>
+      </>
+    );
+  };
+
+  return (
+    <>
+      <div style={inlineStyle}>
+        <div style={{ display: 'flex', alignItems: 'center', height: '82px' }}>
+          {source.name} | {source.type} {renderActionButtons()}
+        </div>
+        {status === 'idle' && canHaveChildFields && isAddFieldBtnDisabled && (
+          <p style={{ fontSize: '12px', margin: '-10px 0 6px', color: '#777' }}>
+            You have reached the maximum depth for the mappings editor. Switch to the{' '}
```

Comment:
I'm guessing `i18n.translate` is WiP for everything.

Personally, I like the current copy here :).